docs/reference/builder.md

```diff
@@ -954,6 +954,125 @@ For example:
+FROM busybox
+ARG user1=
+ARG buildno=
+...
+```
+
+A Dockerfile developer may optionally specify a default value an `ARG` instruction:
+
+```
+FROM busybox
+ARG user1=someuser
+ARG buildno=1
+...
+```
+
+If an `ARG` value has a default and if there is no value passed at build-time, the
+builder uses the default.
+
+An `ARG` variable definition comes into effect from the line on which it is
+defined in the `Dockerfile` not from the argument's use on the command-line or
+elsewhere.  For example, consider this Dockerfile:
+
+```
+1 FROM busybox
+2 USER ${user:-some_user}
+3 ARG user=
+4 USER $user
+...
+```
+A user builds this file by calling:
+
+```
+$ docker build --build-arg user=what_user Dockerfile
+```
+
+The `USER` at line 2 evaluates to `some_user` as the `ARG user` is defined on the
+subsequent line 3. The `USER` at line 4 evaluates to `what_user` as `user` is
+defined and `what_user` was passed on the command line. Prior to its definition by an
+`ARG` instruction, any use of a variable results in an empty string.
+
+> **Note:** It is not recommended to use build-time variables for
+>  passing secrets like github keys, user credentials etc.
+
+You can use an `ARG` or an `ENV` instruction to specify variables that are
+available to the `RUN` instruction. Environment variables defined using the
+`ENV` instruction of `Dockerfile` always override an `ARG` instruction of the same
+name. Consider this Dockerfile with an `ENV` and `ARG` instruction.
+
+```
+1 FROM ubuntu
+2 ARG CONT_IMG_VER=
+3 ENV CONT_IMG_VER v1.0.0
+4 RUN echo $CONT_IMG_VER
+```
+Then, assume this image is built with this command:
+
+```
+$ docker build --build-arg CONT_IMG_VER=v2.0.1 Dockerfile
+```
+
+In this case, the `RUN` instruction uses `v1.0.0` instead of the `ARG` setting
+passed by the user:`v2.0.1` This behavior is similar to a shell
+script where a locally scoped variable overrides the variables passed as
+arguments or inherited from environment, from it's point of definition.
+
+Using the example above but a different `ENV` specification you can create more
+useful interactions between `ARG` and `ENV` instructions:
+
+```
+1 FROM ubuntu
+2 ARG CONT_IMG_VER=
+3 ENV CONT_IMG_VER ${CONT_IMG_VER:-v1.0.0}
+4 RUN echo $CONT_IMG_VER
+```
+
+The command line passes the `--build-arg` and sets the `v2.0.1` value. And the `ARG
+CONT_IMG_VER` is defined on line 2 of the Dockerfile. On line 3, the `ENV`
+instruction of the same name resolves to to `v2.0.1` as the build-time variable
+was passed from command line and expanded here.
+
+The variable expansion technique in this example allows you to pass arguments
+from command-line and persist them in the final image by leveraging the `ENV`
+instruction. Variable expansion is only supported for the `Dockerfile` instructions
+described [here](#environment-replacement).
+
+Unlike an `ARG` instruction, `ENV` values are always persisted in the built image. If
+`docker build` were run without setting the `--build-arg` variable, then
```

Comment:
--build-arg is a flag not an adjective, I believe you mean "build-time variable"